File: xcloak-agent-mobile/test/dfir_test.dart
// Tests for DFIR (digital forensics & incident response) API shapes used by
// the mobile agent. The mobile agent can trigger forensic collection on a
// device and report artifact data back to the backend.
//
// Run with: flutter test

import 'package:flutter_test/flutter_test.dart';

void main() {
  group('DFIR / Forensic Collections (mobile)', () {
    test('forensic collection record includes required status fields', () {
      final col = {
        'id': 1,
        'label': 'Incident 7 Collection',
        'status': 'completed',
        'artifact_types': ['processes', 'connections'],
        'triggered_by': 'soc-analyst',
      };

      for (final key in ['id', 'label', 'status', 'artifact_types']) {
        expect(col.containsKey(key), isTrue,
            reason: 'forensic collection must include "$key" for the DFIR UI');
      }
    });

    test('trigger collection payload uses artifact_types as a list', () {
      final payload = {
        'agent_id': 1,
        'artifact_types': ['processes', 'connections', 'file_hashes'],
      };

      expect(payload['artifact_types'], isA<List>(),
          reason: 'artifact_types must be a list — single string is rejected by backend');
      expect((payload['artifact_types'] as List).isNotEmpty, isTrue);
    });

    test('DFIR URL patterns are correct', () {
      String artifactsUrl(int id) => '/api/dfir/collections/$id/artifacts';
      String timelineUrl(int incidentId) =>
          '/api/dfir/incidents/$incidentId/timeline';

      expect(artifactsUrl(4), equals('/api/dfir/collections/4/artifacts'));
      expect(timelineUrl(7), equals('/api/dfir/incidents/7/timeline'));
    });
  });
}

File: xcloak-agent-desktop/agent/dfir_test.go
package agent

import (
	"encoding/json"
	"fmt"
	"testing"
)

// TestForensicCollection_URLPatterns verifies the agent uses correct URLs for
// DFIR collection operations.
func TestForensicCollection_URLPatterns(t *testing.T) {
	cases := []struct {
		got  string
		want string
	}{
		{fmt.Sprintf("/api/dfir/collections/%d/artifacts", 4), "/api/dfir/collections/4/artifacts"},
		{fmt.Sprintf("/api/dfir/incidents/%d/timeline", 7), "/api/dfir/incidents/7/timeline"},
	}
	for _, c := range cases {
		if c.got != c.want {
			t.Errorf("URL = %q, want %q", c.got, c.want)
		}
	}
}

// TestForensicCollectionPayload_Fields verifies that a forensic collection
// record includes status and label fields used by the frontend.
func TestForensicCollectionPayload_Fields(t *testing.T) {
	raw := `{"id":1,"label":"Incident 7 Collection","status":"completed","artifact_types":["processes","connections"]}`
	var col map[string]any
	if err := json.Unmarshal([]byte(raw), &col); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, key := range []string{"id", "label", "status", "artifact_types"} {
		if _, ok := col[key]; !ok {
			t.Errorf("forensic collection payload missing field %q", key)
		}
	}
}

// TestTriggerCollection_ArtifactTypes verifies that the trigger payload
// accepts an artifact_types array (not a single string).
func TestTriggerCollection_ArtifactTypes(t *testing.T) {
	payload := map[string]any{
		"agent_id":       1,
		"artifact_types": []string{"processes", "connections", "file_hashes"},
	}
	b, _ := json.Marshal(payload)
	var decoded map[string]any
	json.Unmarshal(b, &decoded)
	types, ok := decoded["artifact_types"].([]any)
	if !ok || len(types) == 0 {
		t.Error("artifact_types must be a non-empty array in the trigger payload")
	}
}
